File: cs50x_jorge-manoel_2024/sentimental-credit/credit.py
from cs50 import get_string

number_card = get_string("Number: ")

valid_len = [13, 15, 16]
total_sum = 0

if len(number_card) in valid_len:

    for i in range(len(number_card) - 1, -1, -2):
        total_sum += int(number_card[i])

    for i in range(len(number_card) - 2, -1, -2):
        partial_sum = int(number_card[i]) * 2
        if partial_sum < 10:
            total_sum += partial_sum
        else:
            for n in str(partial_sum):
                total_sum += int(n)

    print(total_sum)

    if total_sum % 10 == 0:
        if number_card[0] == "4" and (len(number_card) == 13 or len(number_card) == 16):
            print("VISA")
        elif int(number_card[0:2]) in range(51, 56) and len(number_card) == 16:
            print("MASTERCARD")
        elif int(number_card[0:2]) in [34, 37] and len(number_card) == 15:
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")

else:
    print("INVALID")

File: cs50x_jorge-manoel_2024/credit_card/credit.c
#include <cs50.h>
#include <stdio.h>

int validate(long card);

long c_number = -1;
int soma = 0;
int i = 1;

int main(void)
{
    while (c_number < 0)
    {
        c_number = get_long("Number: ");
    }

    if (validate(c_number) == 0)
    {
        int id = c_number / 1000000000000;
        if (id / 100 > 50 && id / 100 < 56)
        {
            printf("MASTERCARD\n");
        }
        else if (id / 10 == 34 || id / 10 == 37)
        {
            printf("AMEX\n");
        }
        else if (id / 1000 == 4 || id == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int validate(long card)
{
    while (card > 0)
    {
        int digit = card % 10;

        if (i % 2 == 0)
        {
            digit = digit * 2;
            if (digit > 9)
            {
                digit = digit / 10 + digit % 10;
            }
        }
        soma += digit;
        card = card / 10;
        i++;
    }
    return soma%10;
}
